DSharpPlus/Clients/DiscordClient.Dispatch.cs

```diff
@@ -1163,23 +1171,54 @@ internal async Task OnMessageAckEventAsync(DiscordChannel chn, ulong messageId)
             await this._messageAcknowledged.InvokeAsync(this, new MessageAcknowledgeEventArgs { Message = msg }).ConfigureAwait(false);
         }
 
-        internal async Task OnMessageCreateEventAsync(DiscordMessage message, TransportUser author, TransportMember member)
+        internal async Task OnMessageCreateEventAsync(DiscordMessage message, TransportUser author, TransportMember member, TransportUser referenceAuthor, TransportMember referenceMember)
         {
             message.Discord = this;
+            PopulateMessageReactionsAndCache(message, author, member);
+            PopulateMessageMentions(message, message.Channel?.Guild);
 
             if (message.Channel == null)
                 this.Logger.LogWarning(LoggerEvents.WebSocketReceive, "Channel which the last message belongs to is not in cache - cache state might be invalid!");
             else
                 message.Channel.LastMessageId = message.Id;
 
-            var guild = message.Channel?.Guild;
+            if (message.ReferencedMessage != null)
+            {
+                message.ReferencedMessage.Discord = this;
+                PopulateMessageReactionsAndCache(message.ReferencedMessage, referenceAuthor, referenceMember);
+                PopulateMessageMentions(message, message.ReferencedMessage.Channel?.Guild);
+            }
+
+            var ea = new MessageCreateEventArgs
+            {
+                Message = message,
+
+                MentionedUsers = new ReadOnlyCollection<DiscordUser>(message._mentionedUsers),
+                MentionedRoles = message._mentionedRoles != null ? new ReadOnlyCollection<DiscordRole>(message._mentionedRoles) : null,
+                MentionedChannels = message._mentionedChannels != null ? new ReadOnlyCollection<DiscordChannel>(message._mentionedChannels) : null
+            };
+            await this._messageCreated.InvokeAsync(this, ea).ConfigureAwait(false);
+        }
 
-            this.UpdateMessage(message, author, guild, member);
+        // TODO: I ducking hate this name so much. Halp. 
+        private void PopulateMessageReactionsAndCache(DiscordMessage message, TransportUser author, TransportMember member)
+        {
+            this.UpdateMessage(message, author, message.Channel?.Guild, member);
 
+            if (message._reactions == null)
+                message._reactions = new List<DiscordReaction>();
+            foreach (var xr in message._reactions)
+                xr.Emoji.Discord = this;
+
+            if (this.Configuration.MessageCacheSize > 0 && message.Channel != null)
+                this.MessageCache?.Add(message);
+        }
+
+        private void PopulateMessageMentions(DiscordMessage message, DiscordGuild guild)
+        {
```

Comment:
These methods shouldn't be in the event handler partial class, would probably be best to place it in the main one where `UpdateMessage` is.

Reply:
These were moved to DiscordClient.cs